### ISSUE TITLE
Gevent execmode hangs in Python 3.6.5

### ISSUE THREAD

**Issue by sputnikus** [pytest-dev/execnet#75]
Issue spotted while running pytest-xdist

**Python**: 3.6.5 & 2.7.15
**execnet**: 1.5.0
**gevent**: 1.2.2

Command `pytest -d --tx "2*popen//execmodel=gevent" test.py`

Test file:

```python
from gevent import monkey
monkey.patch_all()


def test_assert_0():
    assert 0


def test_assert_1():
    assert 1
```

In Python 2 environment command runs all tests, reports failure on test_assert_0 and ends. In Python 3 environment test hangs before scanning for tests.

Execnet debug logs for Python 2 & 3 runs https://gist.github.com/sputnikus/13d9ace2bd80ba954b7871dc167acb10

Python 3 workers seem to be working until `[receiver-thread] RECEIVERTHREAD: starting to run` and from my investigation, even `gevent.spawn` gets called, but after that greenlet doesn't run. After that only way is to interrupt run.

Pytest full trace:

```
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! KeyboardInterrupt !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

config = <_pytest.config.Config object at 0x7f1c220ab080>, doit = <function _main at 0x7f1c2553b158>

    def wrap_session(config, doit):
        """Skeleton command line program"""
        session = Session(config)
        session.exitstatus = EXIT_OK
        initstate = 0
        try:
            try:
                config._do_configure()
                initstate = 1
                config.hook.pytest_sessionstart(session=session)
                initstate = 2
>               session.exitstatus = doit(config, session) or 0

/home/sputnikus/.virtualenvs/skypicker_booking_py3/lib/python3.6/site-packages/_pytest/main.py:107: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

config = <_pytest.config.Config object at 0x7f1c220ab080>, session = <Session 'skypicker_booking_py3'>

    def _main(config, session):
        """ default command line protocol for initialization, session,
        running tests and reporting. """
        config.hook.pytest_collection(session=session)
>       config.hook.pytest_runtestloop(session=session)

/home/sputnikus/.virtualenvs/skypicker_booking_py3/lib/python3.6/site-packages/_pytest/main.py:145: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <_HookCaller 'pytest_runtestloop'>, args = (), kwargs = {'session': <Session 'skypicker_booking_py3'>}, notincall = set()

    def __call__(self, *args, **kwargs):
        if args:
            raise TypeError("hook calling supports only keyword arguments")
        assert not self.is_historic()
        if self.argnames:
            notincall = set(self.argnames) - set(['__multicall__']) - set(
                kwargs.keys())
            if notincall:
                warnings.warn(
                    "Argument(s) {} which are declared in the hookspec "
                    "can not be found in this hook call"
                    .format(tuple(notincall)),
                    stacklevel=2,
                )
>       return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)

/home/sputnikus/.virtualenvs/skypicker_booking_py3/lib/python3.6/site-packages/pluggy/__init__.py:617: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <_pytest.config.PytestPluginManager object at 0x7f1c25631cc0>, hook = <_HookCaller 'pytest_runtestloop'>
methods = [<pluggy.HookImpl object at 0x7f1c220abc18>, <pluggy.HookImpl object at 0x7f1c1dfec208>, <pluggy.HookImpl object at 0x7f1c1e217be0>]
kwargs = {'session': <Session 'skypicker_booking_py3'>}

    def _hookexec(self, hook, methods, kwargs):
        # called from all hookcaller instances.
        # enable_tracing will set its own wrapping function at self._inner_hookexec
>       return self._inner_hookexec(hook, methods, kwargs)

/home/sputnikus/.virtualenvs/skypicker_booking_py3/lib/python3.6/site-packages/pluggy/__init__.py:222: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <pluggy._TracedHookExecution object at 0x7f1c1e2b32e8>, hook = <_HookCaller 'pytest_runtestloop'>
hook_impls = [<pluggy.HookImpl object at 0x7f1c220abc18>, <pluggy.HookImpl object at 0x7f1c1dfec208>, <pluggy.HookImpl object at 0x7f1c1e217be0>]
kwargs = {'session': <Session 'skypicker_booking_py3'>}

    def __call__(self, hook, hook_impls, kwargs):
        self.before(hook.name, hook_impls, kwargs)
        outcome = _Result.from_call(lambda: self.oldcall(hook, hook_impls, kwargs))
        self.after(outcome, hook.name, hook_impls, kwargs)
>       return outcome.get_result()

/home/sputnikus/.virtualenvs/skypicker_booking_py3/lib/python3.6/site-packages/pluggy/__init__.py:182: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   outcome = _Result.from_call(lambda: self.oldcall(hook, hook_impls, kwargs))

/home/sputnikus/.virtualenvs/skypicker_booking_py3/lib/python3.6/site-packages/pluggy/__init__.py:180: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

hook = <_HookCaller 'pytest_runtestloop'>
methods = [<pluggy.HookImpl object at 0x7f1c220abc18>, <pluggy.HookImpl object at 0x7f1c1dfec208>, <pluggy.HookImpl object at 0x7f1c1e217be0>]
kwargs = {'session': <Session 'skypicker_booking_py3'>}

    self._inner_hookexec = lambda hook, methods, kwargs: \
        hook.multicall(
            methods, kwargs,
>           firstresult=hook.spec_opts.get('firstresult'),
        )

/home/sputnikus/.virtualenvs/skypicker_booking_py3/lib/python3.6/site-packages/pluggy/__init__.py:216: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <xdist.dsession.DSession object at 0x7f1c1e233b70>

    def pytest_runtestloop(self):
        self.sched = self.config.hook.pytest_xdist_make_scheduler(
            config=self.config,
            log=self.log
        )
        assert self.sched is not None
    
        self.shouldstop = False
        while not self.session_finished:
>           self.loop_once()

/home/sputnikus/.virtualenvs/skypicker_booking_py3/lib/python3.6/site-packages/xdist/dsession.py:116: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <xdist.dsession.DSession object at 0x7f1c1e233b70>

    def loop_once(self):
        """Process one callback from one of the workers."""
        while 1:
            if not self._active_nodes:
                # If everything has died stop looping
                self.triggershutdown()
                raise RuntimeError("Unexpectedly no active workers available")
            try:
>               eventcall = self.queue.get(timeout=2.0)

/home/sputnikus/.virtualenvs/skypicker_booking_py3/lib/python3.6/site-packages/xdist/dsession.py:130: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <queue.Queue object at 0x7f1c1e233ba8>, block = True, timeout = 2.0

    def get(self, block=True, timeout=None):
        '''Remove and return an item from the queue.
    
            If optional args 'block' is true and 'timeout' is None (the default),
            block if necessary until an item is available. If 'timeout' is
            a non-negative number, it blocks at most 'timeout' seconds and raises
            the Empty exception if no item was available within that time.
            Otherwise ('block' is false), return an item if one is immediately
            available, else raise the Empty exception ('timeout' is ignored
            in that case).
            '''
        with self.not_empty:
            if not block:
                if not self._qsize():
                    raise Empty
            elif timeout is None:
                while not self._qsize():
                    self.not_empty.wait()
            elif timeout < 0:
                raise ValueError("'timeout' must be a non-negative number")
            else:
                endtime = time() + timeout
                while not self._qsize():
                    remaining = endtime - time()
                    if remaining <= 0.0:
                        raise Empty
>                   self.not_empty.wait(remaining)

/usr/lib64/python3.6/queue.py:173: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <Condition(<unlocked _thread.lock object at 0x7f1c1e2800a8>, 0)>, timeout = 1.9999989409989212

    def wait(self, timeout=None):
        """Wait until notified or until a timeout occurs.
    
            If the calling thread has not acquired the lock when this method is
            called, a RuntimeError is raised.
    
            This method releases the underlying lock, and then blocks until it is
            awakened by a notify() or notify_all() call for the same condition
            variable in another thread, or until the optional timeout occurs. Once
            awakened or timed out, it re-acquires the lock and returns.
    
            When the timeout argument is present and not None, it should be a
            floating point number specifying a timeout for the operation in seconds
            (or fractions thereof).
    
            When the underlying lock is an RLock, it is not released using its
            release() method, since this may not actually unlock the lock when it
            was acquired multiple times recursively. Instead, an internal interface
            of the RLock class is used, which really unlocks it even when it has
            been recursively acquired several times. Another internal interface is
            then used to restore the recursion level when the lock is reacquired.
    
            """
        if not self._is_owned():
            raise RuntimeError("cannot wait on un-acquired lock")
        waiter = _allocate_lock()
        waiter.acquire()
        self._waiters.append(waiter)
        saved_state = self._release_save()
        gotit = False
        try:    # restore state no matter what (e.g., KeyboardInterrupt)
            if timeout is None:
                waiter.acquire()
                gotit = True
            else:
                if timeout > 0:
>                   gotit = waiter.acquire(True, timeout)
E                   KeyboardInterrupt

/usr/lib64/python3.6/threading.py:299: KeyboardInterrupt
```

**Comment by nicoddemus** [comment]
Thanks @sputnikus!

It would be great if someone finds the time to collaborate a PR for this.

**Comment by ulope** [comment]
The hang also occurs with Python 3.6.4

**Comment by MagicPro1994** [comment]
I fixed this issue by adding timeout to workerpool.waitall() to workerpool.waitall(timeout) on execnet/multi.py.
![image](https://user-images.githubusercontent.com/8509065/61646216-f846a600-acd2-11e9-8c4a-85e0c19b1fc3.png)
I hope it works for you too.

**Comment by RonnyPfannschmidt** [comment]
@MagicPro1994 would you like to submit this as a pr?

**Comment by MagicPro1994** [comment]
@RonnyPfannschmidt I created a pull request. Please review it on ![this PR](https://github.com/pytest-dev/execnet/pull/100)

**Comment by Zac-HD** [comment]
Fixed by #100, though I don't have permissions to close the issue.
